2/answers.txt

```diff
@@ -0,0 +1,231 @@
+making a large number of processes become quickly an expensive activity.
+
+Whereas Nginx uses asynchronous, non-blocking, event driving architecture, allowing for a far more optimal handling of processes, ideally with one process per core, and every thread a subset thereafter.
+
+Q18: What is a page file? What happens to applications that exhaust all available system memory? Is 'paging' bad?
+
+Page files are like caches in that they store data that RAM hasnt used in a while, as a single contiguous chunk of data, so it can be read faster from the HD than from its original location.
+Applications that exhaust all system memory are probably experiencing a leak of some kind, as memory that is no longer required should be shadow dumped and that memory freed up. If however this does happen.
+Usually the program, or the system itself with entirely freeze up and crash, as critical processes are unable to find memory to function correctly.
+
+Paging isnt bad inherently, but poor paging management can be a crutch to a poor design, and when that crutch eventually fails, it will collapse everything.
+
+Q19: Why is dynamic memory, also known as memory ballooning, a useful technology for virtual machines?
+
+For virtual machines, having dynamic memory is important and useful because if a single VM is using less memory than another, memory can be re-allocated or reclaimed to be used by the other vM instances.
+This is particularly useful because it can lower costs than statically ensuring you have enough memory for max output of each VM individualy. but also if you are hosting a set of VMs
+Where the contiguous user or vm usage fluctuates, you can have a smaller stack of memory that can cycle between VMs if the overall max output of all VMs is not higher than the maximum memory available.
+
+Q20: What are processes? Can they share memory?
+
+Processes are instances of programs executing on a core, made up a single thread or multiple threads, there are many forms of IPC (Inter Process Communication) and one of them 
+is the use of shared memory segments, whereby mulitple processes can access the same memory segment. In Unix this is called anonymous shared memory. whereas windows has a native shared memory function.
+Multiple processes cannot share main memory space however, and the creation of share memory segments is a relatively resource expensive task, and should only be used if it would create a larger benefit.
+
+Q21: What are threads? Can they share memory?
+
+Threads are the lighter brother of a process, they are single streams of code execution occuring within the heap, a single thread process will have just one, a multi thread process will obviously have many.
+Each thread has its own stack, but share the heap for the process entire, meaning there is inherently some shared nature of memory within the heap, but the stack itself is isolated.
```

Comment:
Almost.
Processes have a shared memory space for the stack, split between the threads of the process. While each thread has its own section of stack, there's nothing stopping the thread from accessing stack data in another thread of the same process.

It would be an easy way to get a null pointer exception (or dangling pointer) as the threads sharing the data will lose their reference at a non-deterministic point in time when the owning thread completes its current execution and pops the data off its stack.

Due to this, sharing pointers to data residing on the process stack is a bad idea. 

There's a lot more depth to this topic. It's 'low level' systems programming stuff.

On a higher level (which is where we operate for most projects), if a language is garbage collected (c#, java, javascript etc), heap allocations are handled for you and automatically garbage collected (GC) as and when needed. This results in a performance and memory penalty compared to non garbage collected languages. The performance penalty is typically between 1x and 10x for optimised GC languages (c#, java, javascript). It's much more for non-optimised languages like python. This is considered an acceptable tradeoff in the industry for the memory safety and increase in productivity the GC provides.

Some GC languages allow full shared memory access to the heap. 
C# is a language where any usage of async, or spawning a new thread, will allow variables and structs to be used across threads, by default.

The fact that is does this by default means that as a developer you need to be aware of this and use the appropriate thread-safe collections and data synchronisation methods to manipulate data, else you risk shooting yourself in the foot and cause a data race, which is a difficult to debug issue.

Javascript is a GC language that also offers multi-threading but it is far stricter on how it be used. Only raw byte arrays of fixed length can be directly shared between threads, and it's opt-in. Anything else has to be explicitly serialised and sent as a message (technique called message passing). This makes it far easier to implement 'correct' multithreading in javascript, but makes it difficult to process huge datasets in parallel in a performant way.